homeassistant/components/synology_dsm/media_source.py

```diff
@@ -0,0 +1,238 @@
+                        media_class=MediaClass.IMAGE,
+                        media_content_type=mime_type,
+                        title=album_item.file_name,
+                        can_play=True,
+                        can_expand=False,
+                        thumbnail=await self.async_get_thumbnail(
+                            album_item, diskstation
+                        ),
+                    )
+                )
+        return ret
+
+    async def async_resolve_media(self, item: MediaSourceItem) -> PlayMedia:
+        """Resolve media to a url."""
+        identifier = SynologyPhotosMediaSourceIdentifier(item.identifier)
+        if identifier.album_id is None:
+            raise Unresolvable("No album id")
+        if identifier.file_name is None:
+            raise Unresolvable("No file name")
+        mime_type, _ = mimetypes.guess_type(identifier.file_name)
+        if not isinstance(mime_type, str):
+            raise Unresolvable("No file extension")
+        return PlayMedia(
+            f"/synology_dsm/{identifier.unique_id}/{identifier.cache_key}/{identifier.file_name}",
+            mime_type,
+        )
+
+    async def async_get_thumbnail(
+        self, item: SynoPhotosItem, diskstation: SynologyDSMData
+    ) -> str:
+        """Get thumbnail."""
+        try:
+            thumbnail = await diskstation.api.photos.get_item_thumbnail_url(item)
+        except SynologyDSMException:
+            return ""
+        return str(thumbnail)
+
+
+class SynologyDsmMediaView(http.HomeAssistantView):
+    """Synology Media Finder View."""
+
+    url = "/synology_dsm/{source_dir_id}/{location:.*}"
+    name = "synology_dsm"
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the media view."""
+        self.hass = hass
+
+    async def get(
+        self, request: web.Request, source_dir_id: str, location: str
+    ) -> web.FileResponse:
+        """Start a GET request."""
+        if not self.hass.data.get(DOMAIN):
+            raise web.HTTPNotFound()
+        # location: {cache_key}/{filename}
+        cache_key, file_name = location.split("/")
+        image_id = cache_key.split("_")[0]
+        file_extension = file_name.split(".")[-1]
+        mime_type, _ = mimetypes.guess_type(file_name)
+        if not isinstance(mime_type, str):
+            raise web.HTTPNotFound()
+        diskstation: SynologyDSMData = self.hass.data[DOMAIN][source_dir_id]
+
+        item = SynoPhotosItem(image_id, "", "", "", cache_key, "")
+        try:
+            image = await diskstation.api.photos.download_item(item)
+        except SynologyDSMException as exc:
+            raise web.HTTPNotFound() from exc
+        with tempfile.NamedTemporaryFile(
+            mode="wb", suffix=f".{file_extension}", delete=False
+        ) as temp:
+            temp.write(image)
+        return web.FileResponse(Path(temp.name))
```

Comment:
It looks like nothing cleans this up.

Since you are already loading these into memory, it would be better to serve the binary data instead of writing it to disk

Reply:
I changed it from a fileResponse to a normal response